analyzers/src/SonarAnalyzer.Common/Rules/ClassNamedExceptionBase.cs

```diff
@@ -0,0 +1,48 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules;
+
+public abstract class ClassNamedExceptionBase<TSyntaxKind> : SonarDiagnosticAnalyzer<TSyntaxKind>
+    where TSyntaxKind : struct
+{
+    private const string DiagnosticId = "S2166";
+
+    protected abstract TSyntaxKind[] AnalyzedSyntaxKinds { get; }
+    protected override string MessageFormat => "Rename this class to remove \"(e|E)xception\" or correct its inheritance.";
+
+    protected ClassNamedExceptionBase() : base(DiagnosticId) { }
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(
+            Language.GeneratedCodeRecognizer,
+            c =>
+            {
+                if (Language.Syntax.NodeIdentifier(c.Node) is { IsMissing: false } classIdentifier
+                    && classIdentifier.ValueText.EndsWith("Exception", StringComparison.InvariantCultureIgnoreCase)
+                    && c.SemanticModel.GetDeclaredSymbol(c.Node) is INamedTypeSymbol { } classSymbol
+                    && !classSymbol.DerivesFrom(KnownType.System_Exception))
+                {
+                    c.ReportIssue(Diagnostic.Create(Rule, classIdentifier.GetLocation()));
+                }
+            },
+            AnalyzedSyntaxKinds);
```

Comment:
Why are these not in the Language.SyntaxKinds?

Reply:
@zsolt-kolbay-sonarsource Modules are a pain in the @$$ indeed. That being said, the C# equivalent also triggers on the declaration of static classes, so why not the VB.NET one?

After all, if you have a base rule both for C# and VB.NET that does not apply on static classes/modules, you need to deal with that anyway.

Reply:
Fair enough. And we might need in the future for other rules which need both C# and VB.NET implementation.
I've added it to `Language.SyntaxKind` as `ClassAndModuleDeclarations`.